proposals/lip-0049.md

```diff
@@ -223,135 +223,171 @@ When processing a cross-chain message `ccm`, the functions below may be called b
-        and isLive(ccm.receivingChainID):
-        # The Token module is called to transfer the fee from sending to receiving chain
-        token.forwardMessageFee(ccm)
-        if the previous call fails or returns False:
-            return
+def forward(ccu: CCU, ccm: CCM) -> None:
+    # Calculate CCM ID, used later in events.
+    ccmID = SHA-256(encode(crossChainMessageSchema, ccm))
+    
+    # Chain with status REGISTERED do not pass this check.
+    if isLive(ccm.receivingChainID) and chainAccount(ccm.receivingChainID).status == CHAIN_STATUS_ACTIVE:
+        # Transfer the fee from escrow to escrow, call Token LIP
+        Token.transferBetweenEscrows(ccm.sendingChainID, ccm.receivingChainID, feeTokenID, fee)
+
         addToOutbox(ccm.receivingChainID, ccm)
+        # Emit CCM forwarded event.
+        emitEvent(
+            moduleID = MODULE_ID_INTEROPERABILITY,
+            typeID = TYPE_ID_CCM_PROCESSED,
+            data = {"ccmID": ccmID, "status": CCM_PROCESSED_STATUS_FORWARDED},
+            topics = [ccm.sendingChainID, ccm.receivingChainID]
+        )
         return
 
+    # If ccm status is not ok, stop the processing of this message
     if ccm.status != CCM_STATUS_OK:
-        # Stop the processing of this message
+        # Emit CCM discarded event.
+        emitEvent(
+            moduleID = MODULE_ID_INTEROPERABILITY,
+            typeID = TYPE_ID_CCM_PROCESSED,
+            data = {"ccmID": ccmID, "status": CCM_PROCESSED_STATUS_DISCARDED},
+            topics = [ccm.sendingChainID, ccm.receivingChainID]
+        )
         return
 
     # In all following cases, bounce the message with updated status
     # Note: the fee was not assigned to the mainchain relayer and is also bounced
-    bounce(ccm)
+    
+    # Emit CCM processed event: channel unavailable.
+    emitEvent(
+        moduleID = MODULE_ID_INTEROPERABILITY,
+        typeID = TYPE_ID_CCM_PROCESSED,
+        data = {"ccmID": ccmID, "status": CCM_PROCESSED_STATUS_CHANNEL_UNAVAILABLE},
+        topics = [ccm.sendingChainID, ccm.receivingChainID]
+    )
+    ccm.status = CCM_STATUS_CHANNEL_UNAVAILABLE
+    addToOutbox(ccm.sendingChainID, ccm)
 
-    # If the chain does not exist or is not yet active don't continue
-    if (chainAccount(ccm.receivingChainID) does not exist
-        or chainAccount(ccm.receivingChainID).status == CHAIN_REGISTERED):  
+    # If the chain does not exist or is not yet active do not continue
+    if chainAccount(ccm.receivingChainID) does not exist
+        or chainAccount(ccm.receivingChainID).status == CHAIN_STATUS_REGISTERED:  
         return
 
     # In the remaining case, the receiving chain account existed and was either
     # terminated or not live
     # If the receiving chain is still active, then it gets terminated
-    if chainAccount(ccm.receivingChainID).status == CHAIN_ACTIVE:
-        terminateChainInternal(ccm.receivingChainID)
+    if chainAccount(ccm.receivingChainID).status == CHAIN_STATUS_ACTIVE:
```

Comment:
How can the receiving chain be terminated and have `chainAccount(ccm.receivingChainID).status == CHAIN_STATUS_ACTIVE`?

Reply:
I would find the explicit check
```
if not isLive(ccm.receivingChainID) and chainAccount(ccm.receivingChainID).status == CHAIN_STATUS_ACTIVE:
```
here more clear, even if it does a few redundant operations.

Reply:
Sure, notice tho that this explanation is already in the comment.

Reply:
Looks cleaner now.